android/src/main/java/com/vydia/RNUploader/UploadWorker.kt

```diff
@@ -0,0 +1,267 @@
+    // We don't let WorkManager manage retries and network constraints as it's very buggy.
+    // i.e. we'd occasionally get BackgroundServiceStartNotAllowedException,
+    // or ForegroundServiceStartNotAllowedException, or "isStopped" gets set to "true"
+    // for no reason
+    var isRetried = false
+    while (true) {
+      try {
+        // Delay if it's an actual retry.
+        // Should be within the try block to account for worker cancellation,
+        // which resumes immediately and throws CancellationException.
+        if (isRetried) delay(RETRY_DELAY)
+        isRetried = true
+
+        val response = upload() ?: continue
+        handleSuccess(response)
+        return@withContext Result.success()
+      } catch (error: Throwable) {
+        if (checkAndHandleCancellation()) throw error
+        if (checkRetry()) continue
+        handleError(error)
+        throw error
+      }
+    }
+
+    // This should never happen. Only here to satisfy the type check
+    return@withContext Result.failure()
+  }
+
+  private suspend fun upload(): HttpResponse? {
+    val file = File(upload.path)
+    val size = file.length()
+
+    // Register progress asap so the total progress is accurate
+    // This needs to happen before the semaphore wait
+    handleProgress(0, size)
+
+    // Don't bother to run on an invalid network
+    if (!validateAndReportConnectivity()) return null
+
+    // wait for its turn to run
+    semaphore.acquire()
+
+    try {
+      // Use ktor instead of okhttp. Ktor request is coroutine friendly and
+      // will get cancelled when the coroutine gets cancelled
+      var lastProgressReport = 0L
+      val body = file.readChannel()
+      val response = client.request(upload.url) {
+        method = upload.method
+        setBody(body)
+        upload.headers.forEach { (key, value) -> headers.append(key, value) }
+        onUpload { bytesSentTotal, _ ->
+          // throttle progress report
+          val now = System.currentTimeMillis()
+          if (now - lastProgressReport >= PROGRESS_INTERVAL) {
+            lastProgressReport = now
+            handleProgress(bytesSentTotal, size)
+          }
+        }
+      }
+
+      handleProgress(size, size)
+      return response
+    }
+    // don't catch, propagate error up
+    finally {
+      semaphore.release()
+    }
+  }
+
+  private suspend fun handleProgress(bytesSentTotal: Long, fileSize: Long) {
+    UploadProgress.set(context, upload.id, bytesSentTotal, fileSize)
+    EventReporter.progress(upload.id, bytesSentTotal, fileSize)
+    setForeground(getForegroundInfo())
+  }
+
+  private fun handleSuccess(response: HttpResponse) {
+    UploadProgress.scheduleClearing(context)
+    EventReporter.success(upload.id, response)
+  }
+
+  private fun handleError(error: Throwable) {
+    UploadProgress.remove(context, upload.id)
+    UploadProgress.scheduleClearing(context)
+    EventReporter.error(upload.id, error)
```

Comment:
Curious, do these reported errors end up getting stored in Redux state in Diana?

Reply:
We do store the latest error. we also send them to sentry